### DISCUSSION TITLE
ReferenceError: Request is not defined

### DISCUSSION THREAD

**Discussion by AlexanderBachSchmidt** [vercel/next.js#47648] [Help] [answered]
### Summary

Hi,

Suddenly my code is giving this error when deploying on Vercel. I tried tons of things to fix it but nothing seems to work. When I run it locally whether with "next dev", "vercel dev", "vercel build", it doesn't give any error.

What could be causing this and what is the best way to debug it?

### Additional information

```js
ReferenceError: Request is not defined
    at Object.<anonymous> (/vercel/path0/node_modules/next/src/server/web/spec-extension/request.ts:10:34)
    at Module._compile (node:internal/modules/cjs/loader:1165:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1219:10)
    at Object.require.extensions.<computed> [as .js] (/vercel/path0/node_modules/ts-node/src/index.ts:1608:43)
    at Module.load (node:internal/modules/cjs/loader:1043:32)
    at Function.Module._load (node:internal/modules/cjs/loader:878:12)
    at Module.require (node:internal/modules/cjs/loader:1067:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/vercel/path0/node_modules/next/server.js:2:16)
    at Module._compile (node:internal/modules/cjs/loader:1165:14)
```

**Comment by icyJoseph** [comment, marked as answer]
What version of Node.js have you got on your Vercel deployment? If it is 16.x, or less, then you should set it to 18.x, assuming that none of the packages you are ussing, require lower Node.js versions.

Basically, you should have the same Node.js version, locally and on Vercel. 

Figuring out the Node version on Vercel: https://vercel.com/changelog/node-js-version-now-customizable-in-the-project-settings

**Reply by AlexanderBachSchmidt** [reply]
Wow that worked! Thanks so much!!

Just for my understanding, how can it suddenly cause an issue that the node version is wrong? What triggers that it suddenly fails?

**Reply by icyJoseph** [reply]
Node has started to support a few APIs from the Web spec, such as fetch, Request, etc...

So you are locally developing in Node 18, I assume, but your deployment must have been set to earlier versions, say 16.x:

```console
node
Welcome to Node.js v16.14.0.
Type ".help" for more information.
> typeof Request
'undefined'
>
```

While Node 18:

```console
node
Welcome to Node.js v18.14.1.
Type ".help" for more information.
> typeof Request
'function'
>
```

And that's it, there's new APIs in Node 18.

For example, Next.js polyfills fetch with `node-fetch`, but if its present already, it doesn't, and you just use the globally, native, available fetch.

- https://nodejs.org/en/blog/announcements/v18-release-announce#new-globally-available-browser-compatible-apis

**Reply by AlexanderBachSchmidt** [reply]
Thank you! Makes sense!

**Reply by matewilk** [reply]
Thank you, I've come across the same error yesterday and couldn't figure out what was wrong suspecting different packages. You saved the date. What a coincidence!

Much appreciated!

**Reply by faical23** [reply]
i get the some error and i fix it with change my current node js from 16.15 to 18 , thank you

**Reply by GabrielDelight** [reply]
Thanks a lot

**Reply by anyb1s** [reply]
Regardless which version I run, still get this, 
ReferenceError: Request is not defined
    at Object.<anonymous> (node_modules/next/dist/server/web/spec-extension/request.js:28:27)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Module._load (node:internal/modules/cjs/loader:1024:12)
    at Module.require (node:internal/modules/cjs/loader:1233:19)
    at mod.require (/Users/galindenev/Sites/agency/node_modules/next/dist/server/require-hook.js:65:28)
    at require (node:internal/modules/helpers:179:18)
    at Object.<anonymous> (/Users/galindenev/Sites/agency/node_modules/next/dist/server/web/spec-extension/adapters/next-request.js:37:18)
    at Module._compile (node:internal/modules/cjs/loader:1358:14)

Node.js v20.16.0

**Comment by jvmhonorato** [comment]
To solve this on Windows: download the latest LTS version: https://nodejs.org/en/download


To solve this in WSL 2 (Linux):

Step1:  First, let’s install Node Version Manager (nvm), which will enable us to manage Node.js versions easily.
 ```
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.38.0/install.sh | bash

```

Step2:  With nvm installed, you can now check the available Node.js versions:

```
nvm ls-remote

```
Step3: Now you will install the latest LTS version of Node in this case in 10/2023 it is 20.9:
```
nvm install 20.09

```

Step4: Just to be sure, put the installed version as default

```
nvm alias default 20.9

```

Step5: Check the Node version:

```
node -v

```

Step6: Be happy and nice coding

**Reply by sunkanmii** [reply]
Yep, it works for Windows. Was on `v16.17.0` before, installed the latest version (`v20.9.0`) from the node.js site and it works!
Thanks!

**Comment by GabrielDelight** [comment]
[Fixed] Just update your node js version 

nodejs -v // version 16 😢 
sudo npm install -g n // Install the n package 
sudo n lts    // Update to lts version 
nodejs -v // version 20 😄

**Reply by monkeyK1n9** [reply]
This worked for me. Thanks.

I am on a Mac M1
Checking my node version:```node -v ```, I was at v16.20.1, I then upgraded my node version using **nvm** to the LTS version.
```nvm install --lts```. At the moment of writing this comment, I am now at v20.9.0.

Everything works now fine.

**Reply by GabrielDelight** [reply]
I'm glad my answer was able to help xd

**Reply by pedrotomas50** [reply]
Its works but the old projects doesnt run dev.

**Reply by jvmhonorato** [reply]
go to package.json  file in ur old project and update the NODE. exp: "node":"16.0.9" to "node":"20.0.9", afterwards type "npm install" on terminal.

**Comment by vktechturbine** [comment]
i face with same problem i have solution
first check current node version in your system  
then first update to latest  node version  using this command 
in terminal and type =>
     1. node -v => check node version
     2. nvm ls-remote ->  is in list to check latest node version copy version number v20.0.0
     3. nvm ls => is use to how many version you have in your system(optional)
     4. nvm install  v20.9.0 =>  this command to install latest version in your system
     5. nvm use v20.9.0 -> this command is to use to show now ure useing this version
     6. same procedure with your project terminal also then solve the error

**Reply by muftisamiullah** [reply]
doesnt seems to work for me, On mac, nextjs version 14.0.0 or with 14.0.1, nd tried above solutions.

**Reply by ishaanbuildsthings** [reply]
This also does not work for me:

`TypeError: Cannot convert undefined or null to object`

**Reply by icyJoseph** [reply]
mmmm this can be that for some reason Request, part of the Fetch API family, is not defined for some reason in your environment.

Could you try to create a script (forget about next.js for now), say `scripts/index.js` and run it with `node script/index.js`:

```js
const request = new Request("https://www.mozilla.org/favicon.ico");

const url = request.url;
const method = request.method;
const credentials = request.credentials;

console.log({ url, method, credentials })
```

Does this work for you?

**Reply by RomneyDa** [reply]
When I ran this script it didn't work, same error

**Comment by rameezjafrie** [comment]
Many Thanks after updating node version, issue resolved.

**Comment by moonRider** [comment]
<img width="346" alt="image" src="https://github.com/vercel/next.js/assets/6049820/2546a97a-ce1c-484f-8a39-fbc6d078d8c7">
but it works with `sudo`
<img width="420" alt="image" src="https://github.com/vercel/next.js/assets/6049820/7245ea97-ad13-4c8d-8624-7ef3e7c7db45">

**Reply by moonRider** [reply]
I solved this problem by remove NODE_OPTIONS.
